apps/arweave/src/ar_tx.erl

```diff
@@ -440,21 +454,42 @@ ends_with_digit(Data) ->
 	LastByte = binary:last(Data),
 	LastByte >= 48 andalso LastByte =< 57.
 
+verify_signature_type(#tx { format = Format, signature_type = TXSigType }, Height) ->
+	case Height < ar_fork:height_2_5() of
+		true -> Format > 1;
```

Comment:
Should `ar_serialize:json_struct_to_tx/1` return a dummy signature type so we can catch invalid `signature_type` values here?

Reply:
signature_type=undefined essentially means RSA now so I would set the signature type right away on deserialization (or as a #tx record default if the tx is created internally).

Regarding `Format > 1;`, we are not dropping support for v1 transactions in 2.5, we just need to assert the type is one of the three supported types.

Reply:
> signature_type=undefined essentially means RSA now so I would set the signature type right away on deserialization (or as a #tx record default if the tx is created internally).
I don't believe we can do this as we need to know whether or not `signature_type` was defined when the tx was signed so we can check its signature appropriately.

> Regarding Format > 1;, we are not dropping support for v1 transactions in 2.5
Oops, yeah. Will fix that.

> we just need to assert the type is one of the three supported types.
From what I can tell, right now when someone sends in a tx with an invalid `signature_type` we'll get an error in `ar_serialize:json_struct_to_tx/1` as none of the case conditions will match. That was why I was wondering if it would make sense to return a dummy value like an `invalid_signature` atom when deserializing so `ar_tx:verify_signature_type/2` can validate things appropriately. What do you think?

Reply:
We need to ensure compatibility with the existing clients; therefore, if there is no `signature_type` field we need to consider it an RSA-signed transaction.

Hm, if we detect an invalid signature in `json_struct_to_tx` already, why pass it on? Obviously, `ar_tx:verify` must in any case fail when facing an invalid signature type, but it can do it miserably (e.g., because `case SignatureType of` does not have a clause for the specified unsupported type).

Reply:
> We need to ensure compatibility with the existing clients; therefore, if there is no signature_type field we need to consider it an RSA-signed transaction.

Yeah, definitely. Currently, I am doing this by checking whether or not `signature_type` is defined and using `?DEFAULT_KEY_TYPE` when verifying if it isn't. Here's an [example](https://github.com/ArweaveTeam/arweave/pull/309/files#diff-a7235baf70043c993ba2c2a7dcaab49f31756d54297f0fb6703f98d9cf43fcb0R465).

I do this as opposed to just defaulting `signature_type` on txs because `signature_data_segment_v2/1` needs a way to know whether or not `signature_type` was defined in the original tx to generate the right signature data [here](https://github.com/ArweaveTeam/arweave/pull/309/files#diff-a7235baf70043c993ba2c2a7dcaab49f31756d54297f0fb6703f98d9cf43fcb0R219).

As an alternative to this we can maybe include a boolean on the internal tx representation on whether or not `signature_type` was included when deserializing but I don't think this is as elegant.

Reply:
> Hm, if we detect an invalid signature in json_struct_to_tx already, why pass it on? Obviously, ar_tx:verify must in any case fail when facing an invalid signature type, but it can do it miserably (e.g., because case SignatureType of does not have a clause for the specified unsupported type).

If we do this users won't get an `invalid_signature_type` error from the HTTP API when they specify an invalid `signature_type` right? It isn't a big issue, I was just wondering if we should do this for dev API ergonomics sake.

Reply:
> I do this as opposed to just defaulting signature_type on txs because signature_data_segment_v2/1 needs a way to know whether or not signature_type was defined in the original tx to generate the right signature data.

There is another alternative - do not include the signature type in the segment of RSA transactions. I like it more because otherwise we get two redundant ways of constructing and validating an RSA transaction; moreover, the "new" format (with the type explicitly specified) cannot be verified by the existing clients. 

> If we do this users won't get an invalid_signature_type error from the HTTP API when they specify an invalid signature_type right

Good point! We could update [the handler](https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_http_iface_middleware.erl#L1932) accordingly and [record the error code](https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_tx.erl#L312) in the serialization function, WDYT? The error cache is used by the `GET /tx/[id]` handler to display the failure reason for recently failed transactions.

Reply:
> There is another alternative - do not include the signature type in the segment of RSA transactions. I like it more because otherwise we get two redundant ways of constructing and validating an RSA transaction;

Ahh, I like this method too since it removes the need for all the `undefined` checks.

> We could update the handler accordingly and record the error code in the serialization function, WDYT? The error cache is used by the GET /tx/[id] handler to display the failure reason for recently failed transactions.

Sounds good. Have a look at my recent commit, I think I got what you meant.